qutip/core/qobj.py

```diff
@@ -206,503 +270,217 @@ class Qobj(object):
-            out.dims = self.dims
-            out.superrep = self.superrep
-            if settings.auto_tidyup: out.tidyup()
-            if isinstance(other, complex):
-                out._isherm = out.isherm
-            else:
-                out._isherm = self._isherm
-
-            return out
+        # Shouldn't be here unless `other.__mul__` has already been tried, so
+        # we _shouldn't_ check that `other` is `Qobj`.
+        if not isinstance(other, numbers.Number):
+            return NotImplemented
+        return self.__mul__(complex(other))
 
-        else:
-            raise TypeError("Incompatible object for multiplication")
+    def __matmul__(self, other):
+        if not isinstance(other, Qobj):
+            try:
+                other = Qobj(other)
+            except TypeError:
+                return NotImplemented
+        if self.dims[1] != other.dims[0]:
+            raise TypeError("".join([
+                "incompatible dimensions ",
+                repr(self.dims),
+                " and ",
+                repr(other.dims),
+            ]))
+        if (
+            self.type == 'super'
+            and other.type == 'super'
+            and self.superrep != other.superrep
+        ):
+            raise TypeError("".join([
+                "incompatible superoperator representations ",
+                repr(self.superrep),
+                " and ",
+                repr(other.superrep),
+            ]))
+        if self.type == 'bra' and other.type == 'ket':
+            return _data.inner_csr(self.data, other.data)
+        return Qobj(_data.matmul_csr(self.data, other.data),
+                    dims=[self.dims[0], other.dims[1]],
+                    isherm=self._isherm and other._isherm,
+                    isunitary=self._isunitary and other._isunitary,
+                    superrep=self.superrep,
+                    copy=False)
 
     def __truediv__(self, other):
-        return self.__div__(other)
-
-    def __div__(self, other):
-        """
-        DIVISION (by numbers only)
-        """
-        if isinstance(other, Qobj):  # if both are quantum objects
-            raise TypeError("Incompatible Qobj shapes " +
-                            "[division with Qobj not implemented]")
-
-        if isinstance(other, (int, float, complex,
-                              np.integer, np.floating, np.complexfloating)):
-            out = Qobj()
-            out.data = self.data / other
-            out.dims = self.dims
-            if settings.auto_tidyup: out.tidyup()
-            if isinstance(other, complex):
-                out._isherm = out.isherm
-            else:
-                out._isherm = self._isherm
-
-            out.superrep = self.superrep
-
-            return out
-
-        else:
-            raise TypeError("Incompatible object for division")
+        if not isinstance(other, numbers.Number):
+            return NotImplemented
+        return self.__mul__(complex(other))
```

Comment:
Whoops: should be `self.__mul__(1 / complex(other))`

Reply:
Ignoring of course that _technically_ in floating point those aren't necessarily equivalent operations.